### PR TITLE
Sort by combination of cell suspension and project cell counts (#3581)

### PR DESCRIPTION
**Pull request by danielsotirhos**
https://github.com/DataBiosphere/azul/issues/3581

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved issues to `prod` or `Merged prod`               <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                     <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Review by amarjandu — APPROVED** [review]

**Comment by amarjandu** [comment]
#3651 merged, chain shortened

File: test/service/test_response.py [hunk 1]
```diff
@@ -2607,76 +2618,112 @@ def tearDownClass(cls):
...
+                with self.subTest(facet=facet, ascending=ascending):
+                    params = {
+                        'catalog': self.catalog,
+                        'sort': facet,
+                        'order': 'asc' if ascending else 'desc'
+                    }
+                    url = self.base_url.set(path='/index/projects', args=params)
+                    response = requests.get(str(url))
+                    response.raise_for_status()
+                    response_json = response.json()
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2672]:
```suggestion
                    response = response.json()
```

File: test/service/test_response.py [hunk 2]
```diff
@@ -2607,76 +2618,112 @@ def tearDownClass(cls):
...
+                    actual = self._summarize_cell_counts(response_json)
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2673]:
```suggestion
                    actual = self.CellCounts.from_response(response)
```

File: test/service/test_response.py [hunk 3]
```diff
@@ -2607,76 +2618,112 @@ def tearDownClass(cls):
...
-            (1, 349),
-            (1, 6210),
-            (2, 7738),
-            (1, 10000)
-        ]
-        for ascending in (True, False):
-            with self.subTest(ascending=ascending):
+        test_data = {
+            'cellCount': [
+                {'cell_suspensions': 1, 'cell_suspension_cells': 0, 'project_cells': 88_000},
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2640]:
```suggestion
                self.CellCounts(88_000, heart=0),
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/service/test_response.py [hunk 4]
```diff
@@ -2613,76 +2624,125 @@ def tearDownClass(cls):
...
-        expected_results = [
-            (1, 6210),
-            (2, 7738)
-        ]
-        self.assertEqual(actual_results, expected_results)
+                response = response.json()
+                actual = self.CellCounts.from_response(response)
+                self.assertEqual(actual, filter_data['results'])
+
+    class CellCounts:
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2722]: Should probably go before the methods.

File: test/service/test_response.py [hunk 5]
```diff
@@ -2613,76 +2624,125 @@ def tearDownClass(cls):
...
+
+        def __init__(self, project_cells: Optional[int], **kwargs):
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2724]:
```suggestion
        def __init__(self, project_cells: Optional[int], cells_by_organ):
```

File: test/service/test_response.py [hunk 6]
```diff
@@ -2613,76 +2624,125 @@ def tearDownClass(cls):
...
-        return [
-            (
-                len(hit['cellSuspensions']),
-                sum([cs['totalCells'] for cs in hit['cellSuspensions']])
-            )
-            for hit in response_json['hits']
-        ]
+        test_data = {
+            'cellCount': [
+                self.CellCounts(88_000, **{'mouth mucosa': None}),
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2634]:
```suggestion
                self.CellCounts(88_000, {'mouth mucosa': None}),
```

File: test/service/test_response.py [hunk 7]
```diff
@@ -2613,76 +2624,125 @@ def tearDownClass(cls):
...
-        self.assertEqual(actual_results, expected_results)
+                response = response.json()
+                actual = self.CellCounts.from_response(response)
+                self.assertEqual(actual, filter_data['results'])
+
+    class CellCounts:
+
+        def __init__(self, project_cells: Optional[int], **kwargs):
+            self.project_cells = project_cells
+            self.__dict__.update(kwargs)
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2726]: I did not mean to imply that each keyword arg becomes an attribute. Since organs have spaces, we should probably ditch the keyword arg approach and just use a attrs class.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/metadata/hca/aggregate.py [hunk 8]
```diff
@@ -47,20 +51,31 @@
...
+            project_cells = project['estimated_cell_count'] or 0
+            return max(project_cells, self.cell_count)
+        else:
+            return self.cell_count
+
     @classmethod
     def field_types(cls, field_types: FieldTypes) -> FieldTypes:
         return dict(super().field_types(field_types),
-                    total_estimated_cells=pass_thru_int)
+                    cell_count=pass_thru_int,
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/aggregate.py:72]: I don't think this is the right field type. Your test demonstrates that both of these can be None and we don't use pass through types unless we are sure they that either can't be None or we don't need to filter by them.

File: test/service/test_response.py [hunk 9]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
...
 
-    def _count_total_cells(self, response_json):
+    @attr.s(auto_attribs=True, frozen=True)
+    class CellCounts:
+        project_cells: int
+        cells_by_organ: Dict[str, int]
+
+        @classmethod
+        def from_response(cls, response: JSON) -> List['TestSortAndFilterByCellCount.CellCounts']:
+            return [
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2635]: The list comprehension should not be a concern of this factory.

File: test/service/test_response.py [hunk 10]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
         cls._teardown_indices()
         super().tearDownClass()
 
-    def _count_total_cells(self, response_json):
+    @attr.s(auto_attribs=True, frozen=True)
+    class CellCounts:
+        project_cells: int
+        cells_by_organ: Dict[str, int]
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2631]: They type hint doesn't reflect the optionality. PyCHarm shows me a warning about this.

File: test/service/test_response.py [hunk 11]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
...
+                    }
+                    url = self.base_url.set(path='/index/projects', args=params)
+                    response = requests.get(str(url))
+                    response.raise_for_status()
+                    response = response.json()
+                    actual = self.CellCounts.from_response(response)
+                    if ascending:
+                        expected = test_data[facet]
+                    else:
+                        expected = list(reversed(test_data[facet]))
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2689]: A little verbose

File: test/service/test_response.py [hunk 12]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
...
-        ascending_results = [
-            (1, 1),
-            (1, 349),
-            (1, 6210),
-            (2, 7738),
-            (1, 10000)
-        ]
-        for ascending in (True, False):
-            with self.subTest(ascending=ascending):
+        test_data = {
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2697]: Everything is data

File: test/service/test_response.py [hunk 13]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
...
+            },
+            'effectiveCellCount': {
+                'within': [3000, 7000],
+                'results': [
+                    self.CellCounts(3589, {'brain': 1}),
+                    self.CellCounts(None, {'Brain': 6210}),
+                ],
+            },
+        }
+        for facet, filter_data in test_data.items():
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2719]: Everything is data, we have better, more descriptive names for the elements of a filter.

Comment (hannes-ucsc) [reply test/service/test_response.py:2719]: Facets are fields that occur in the termFacets part of the response.

File: test/service/test_response.py [hunk 14]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
...
+                ],
+            },
+            'projectEstimatedCellCount': {
+                'within': [6000, 100_000],
+                'results': [
+                    self.CellCounts(88_000, {'mouth mucosa': None}),
+                ],
+            },
+            'effectiveCellCount': {
+                'within': [3000, 7000],
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2712]: Natively, ranges are tuples, not lists, we just can't express that distinction in JSON. We can here.

Comment (hannes-ucsc) [reply test/service/test_response.py:2712]: When the definition of the fingerprint (or digest) is so close to its use, it can be a little more compact.

File: test/service/test_response.py [hunk 15]
```diff
@@ -2613,76 +2625,117 @@ def tearDownClass(cls):
...
-        response_json = response.json()
-        actual_results = self._count_total_cells(response_json)
-        expected_results = [
-            (1, 6210),
-            (2, 7738)
-        ]
-        self.assertEqual(actual_results, expected_results)
+                response = response.json()
+                actual = self.CellCounts.from_response(response)
+                self.assertEqual(actual, filter_data['results'])
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2738]: Expectation always first

Comment (hannes-ucsc) [reply test/service/test_response.py:2738]: It seems we need a test for filtering by a `null` value.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: test/service/test_response.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_response.py b/test/service/test_response.py
--- a/test/service/test_response.py	(revision 49c1f2a8c0ca6cce2ae56e29eabe161cf0782428)
+++ b/test/service/test_response.py	(date 1644453958336)
@@ -2582,6 +2582,20 @@
         ])
 
 
+@attr.s(auto_attribs=True, frozen=True)
+class CellCounts:
+    estimated_cell_count: Optional[int]
+    total_cells: Dict[str, Optional[int]]
+
+    @classmethod
+    def from_response(cls, hit: JSON) -> 'CellCounts':
+        return cls(estimated_cell_count=one(hit['projects'])['estimatedCellCount'],
+                   total_cells={
+                       one(cell_suspension['organ']): cell_suspension['totalCells']
+                       for cell_suspension in hit['cellSuspensions']
+                   })
+
+
 @patch_dss_endpoint
 @patch_source_cache
 class TestSortAndFilterByCellCount(WebServiceTestCase):
@@ -2625,117 +2639,92 @@
         cls._teardown_indices()
         super().tearDownClass()
 
-    @attr.s(auto_attribs=True, frozen=True)
-    class CellCounts:
-        project_cells: int
-        cells_by_organ: Dict[str, int]
-
-        @classmethod
-        def from_response(cls, response: JSON) -> List['TestSortAndFilterByCellCount.CellCounts']:
-            return [
-                cls(one(hit['projects'])['estimatedCellCount'],
-                    {
-                        one(cs['organ']): cs['totalCells']
-                        for cs in hit['cellSuspensions']
-                    }
-                    )
-                for hit in response['hits']
-            ]
-
     def test_sorting_by_cell_count(self):
         """
         Verify sorting projects by the total cell suspension cell count, the
         project cell count, and the effective cell count.
         """
-        test_data = {
+        test_cases = {
             'cellCount': [
-                self.CellCounts(88_000, {'mouth mucosa': None}),
-                self.CellCounts(3589, {'brain': 1}),
-                self.CellCounts(None, {'Brain': 6210}),
-                self.CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
-                self.CellCounts(None, {'brain': 10_000}),
+                CellCounts(88_000, {'mouth mucosa': None}),
+                CellCounts(3589, {'brain': 1}),
+                CellCounts(None, {'Brain': 6210}),
+                CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
+                CellCounts(None, {'brain': 10_000}),
             ],
             'projectEstimatedCellCount': [
-                self.CellCounts(3589, {'brain': 1}),
-                self.CellCounts(88_000, {'mouth mucosa': None}),
-                self.CellCounts(None, {'Brain': 6210}),
-                self.CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
-                self.CellCounts(None, {'brain': 10_000}),
+                CellCounts(3589, {'brain': 1}),
+                CellCounts(88_000, {'mouth mucosa': None}),
+                CellCounts(None, {'Brain': 6210}),
+                CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
+                CellCounts(None, {'brain': 10_000}),
             ],
             'effectiveCellCount': [
-                self.CellCounts(3589, {'brain': 1}),
-                self.CellCounts(None, {'Brain': 6210}),
-                self.CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
-                self.CellCounts(None, {'brain': 10_000}),
-                self.CellCounts(88_000, {'mouth mucosa': None}),
+                CellCounts(3589, {'brain': 1}),
+                CellCounts(None, {'Brain': 6210}),
+                CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
+                CellCounts(None, {'brain': 10_000}),
+                CellCounts(88_000, {'mouth mucosa': None}),
             ]
         }
-        for facet in test_data:
-            for ascending in (True, False):
-                with self.subTest(facet=facet, ascending=ascending):
+        for ascending in False, True:
+            for field, expected in test_cases.items():
+                with self.subTest(facet=field, ascending=ascending):
                     params = {
                         'catalog': self.catalog,
-                        'sort': facet,
+                        'sort': field,
                         'order': 'asc' if ascending else 'desc'
                     }
                     url = self.base_url.set(path='/index/projects', args=params)
                     response = requests.get(str(url))
                     response.raise_for_status()
                     response = response.json()
-                    actual = self.CellCounts.from_response(response)
-                    if ascending:
-                        expected = test_data[facet]
-                    else:
-                        expected = list(reversed(test_data[facet]))
-                    self.assertEqual(actual, expected)
+                    actual = list(map(CellCounts.from_response, response['hits']))
+                    if not ascending:
+                        expected = list(reversed(expected))
+                    self.assertEqual(expected, actual)
 
     def test_filter_by_cell_count(self):
         """
         Verify filtering projects by the total cell suspension cell count, the
         project cell count, and the effective cell count.
         """
-        test_data = {
-            'cellCount': {
-                'within': [3000, 8000],
-                'results': [
-                    self.CellCounts(None, {'Brain': 6210}),
-                    self.CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
+        test_cases = {
+            'cellCount': (
+                (3000, 8000),
+                [
+                    CellCounts(None, {'Brain': 6210}),
+                    CellCounts(None, {'presumptive gut': 3869, 'endoderm': 3869}),
                 ],
-            },
-            'projectEstimatedCellCount': {
-                'within': [6000, 100_000],
-                'results': [
-                    self.CellCounts(88_000, {'mouth mucosa': None}),
+            ),
+            'projectEstimatedCellCount': (
+                (6000, 100_000),
+                [
+                    CellCounts(88_000, {'mouth mucosa': None}),
                 ],
-            },
-            'effectiveCellCount': {
-                'within': [3000, 7000],
-                'results': [
-                    self.CellCounts(3589, {'brain': 1}),
-                    self.CellCounts(None, {'Brain': 6210}),
+            ),
+            'effectiveCellCount': (
+                (3000, 7000),
+                [
+                    CellCounts(3589, {'brain': 1}),
+                    CellCounts(None, {'Brain': 6210}),
                 ],
-            },
+            ),
         }
-        for facet, filter_data in test_data.items():
-            with self.subTest(facet=facet):
+        for field, (within, expected) in test_cases.items():
+            with self.subTest(facet=field):
                 params = {
                     'catalog': self.catalog,
-                    'sort': facet,
+                    'sort': field,
                     'order': 'asc',
-                    'filters': json.dumps({
-                        facet: {
-                            'within': [
-                                filter_data['within']
-                            ]
-                        }
-                    })
+                    'filters': json.dumps({field: {'within': [within]}})
                 }
                 url = self.base_url.set(path='/index/projects', args=params)
                 response = requests.get(str(url))
                 response.raise_for_status()
                 response = response.json()
-                actual = self.CellCounts.from_response(response)
-                self.assertEqual(actual, filter_data['results'])
+                actual = list(map(CellCounts.from_response, response['hits']))
+                self.assertEqual(actual, expected)
 
 
 @patch_dss_endpoint
```

File: test/service/test_response.py [hunk 16]
```diff
@@ -2654,76 +2680,110 @@ def tearDownClass(cls):
...
-        response.raise_for_status()
-        response_json = response.json()
-        actual_results = self._count_total_cells(response_json)
-        expected_results = [
-            (1, 6210),
-            (2, 7738)
-        ]
-        self.assertEqual(actual_results, expected_results)
+        for field, test_case in test_cases.items():
+            for value, expected in test_case.items():
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2769]:
```suggestion
            for filter, expected in test_case.items():
```

File: test/service/test_response.py [hunk 17]
```diff
@@ -2654,76 +2680,110 @@ def tearDownClass(cls):
...
+                with self.subTest(facet=field, value=value):
+                    if isinstance(value, tuple):
+                        filters = {field: {'within': [value]}}
+                    else:
+                        filters = {field: {'is': [value]}}
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2774]:
```suggestion
filters = {
     field: {'within': [value]} 
            if isinstance(value, tuple) else 
            {'is': [value]}
}
```

File: test/service/test_response.py [hunk 18]
```diff
@@ -2654,76 +2680,110 @@ def tearDownClass(cls):
...
-                    'order': 'asc' if ascending else 'desc'
-                }
-                url = self.base_url.set(path='/index/projects', args=params)
-                response = requests.get(str(url))
-                response.raise_for_status()
-                response_json = response.json()
-                actual_results = self._count_total_cells(response_json)
-                expected = ascending_results if ascending else list(reversed(ascending_results))
-                self.assertEqual(actual_results, expected)
+        test_cases = {
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2688]: Did you have to adjust the expectations to accommodate the change in field type?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/service/test_response.py (hunk 18 above)
Comment (danielsotirhos) [reply test/service/test_response.py:2688]: The change in field type (`pass_thru_int` to `null_int`) did not affect the results in the test of sorting, however the change was required in order to filter by a cell count of `None`.

Comment (hannes-ucsc) [reply test/service/test_response.py:2688]: Makes sense.

File: test/service/test_response.py [hunk 19]
```diff
@@ -2654,76 +2680,112 @@ def tearDownClass(cls):
...
-        response.raise_for_status()
-        response_json = response.json()
-        actual_results = self._count_total_cells(response_json)
-        expected_results = [
-            (1, 6210),
-            (2, 7738)
-        ]
-        self.assertEqual(actual_results, expected_results)
+        for field, test_case in test_cases.items():
+            for filter_, expected in test_case.items():
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2769]:
```suggestion
            for filter, expected in test_case.items():
```

If you get a warning about shadowing a built-in, disable that inspection globally.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — APPROVED** [review]
